### ISSUE TITLE
I think maybe there is a very rare bug in map

### ISSUE THREAD

**Issue by xzores** [odin-lang/Odin#2667]
## Context

I am using map to index a chunk for my game. The index I use is a [3]i32.
Now it seems there there is a very rare chance that the map cannot delete a key.
very rare is like 1 out of 100K to 1M times.

Info:
Odin: dev-2023-07-nightly:3072479c
OS:   Windows 11 Home Basic (version: 22H2), build 22621.1992
CPU:  13th Gen Intel(R) Core(TM) i9-13900HX
RAM:  32507 MiB

## Expected Behavior

I would expect delete_key to remove the element with the given key from the map.

## Current Behavior

It very rarely does not delete the key.

It simply tries to remove the map every frame, but it seems to fail for some reason, my logs:
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]
removing : [-28, 8, 25]

## Failure Information (for bugs)

Ok, since it is a rare phenomenon I will give you all the code I use
```
//This ordering matter
Chunk_state :: enum {
    should_remove,      //This shall be destroyed.
    is_passive,         //Passive is when it is close enough that is should not be destroyed, but not close enough to be generated.
    is_active,          //is_active is when it is close enough that is should generated if it was not.
    generation_ungoing,  //The generation has begun, but is not done yet, and so it shall not be deleted or generated again.
    ungenerated,        //Tells that the chuck has just been created this frame and it needs to be generated.
}

Chunk :: struct {
    blocks : [5][5][5]Block,
    chunk_state : Chunk_state,
}

```

```
Chunk_size_blocks : i32 : 256;
Chunk_size_world : f32 : 4;

to_chunk_space :: proc(pos : [3]f32) -> [3]i32 {
    return {cast(i32)(pos.x / Chunk_size_world), cast(i32)(pos.y / Chunk_size_world), cast(i32)(pos.z / Chunk_size_world)} ;
}

to_world_space :: proc(ipos : [3]i32) -> [3]f32 {
    return cast(f32)Chunk_size_world * [3]f32{cast(f32)ipos.x,  cast(f32)ipos.y, cast(f32)ipos.z};
}

Terrain :: struct {
    chunks : map[[3]i32]Chunk,
    agents : [dynamic][3]f32,

    render_distance : f32,      //in chunks
    unrender_distance : f32,    //in chunks
}

init :: proc () -> Terrain {

    using res : Terrain;

    for ipos, &c in chunks {
        _init_chunk(&c);
    }
    
    render_distance = 5;
    unrender_distance = 8;

    return res;
}

update :: proc(using terrain : ^Terrain) {
    
    Gen_info :: struct {
        ipos : [3]i32,
        state : Chunk_state,
    }

    for _, &c in chunks {
        if c.chunk_state < .generation_ungoing { //don't delete ungenerated chunks.
            c.chunk_state = .should_remove;
        }
    }

    for apos in agents {
        ipos := to_chunk_space(apos);

        max_check : i32 = 20;

        for x : i32 = ipos.x - max_check; x < ipos.x + max_check ; x+=1 {
            for y : i32 = ipos.y - max_check; y < ipos.y + max_check ; y+=1 {
                for z : i32 = ipos.z - max_check; z < ipos.z + max_check ; z+=1 {
                    
                    //TODO this distance is incorrect? should we do it in world space, that would be cool.
                    distance : f32 = linalg.distance(apos, to_world_space({x, y, z})); //distance is chunk space.
                    new_state : Chunk_state = .should_remove;

                    if distance < render_distance * Chunk_size_world {
                        new_state = .is_active;
                    }
                    else if distance < unrender_distance * Chunk_size_world {
                        new_state = .is_passive;
                    }
                    
                    if new_state != .should_remove {
                        if !({x,y,z} in chunks) && new_state == .is_active {
                            //fmt.printf("adding : %v\n", [3]i32{x,y,z});
                            chunks[{x,y,z}] = Chunk{};
                            new_state = .ungenerated;
                        }
                    }

                    if ([3]i32{x,y,z} in chunks) {
                        c : ^Chunk = &chunks[{x,y,z}]
                        //fmt.printf("accessing : %v\n", [3]i32{x,y,z});
                        if new_state > c.chunk_state {
                            c.chunk_state = new_state;
                        }
                    }
                }
            }
        }
    }
    
    to_remove : [dynamic][3]i32 = make([dynamic][3]i32);
    defer delete(to_remove);

    for ipos, c in &chunks {
        if c.chunk_state == .ungenerated && rnd.float64_range(0, 100) < 1 {
            c.chunk_state = .is_passive;
        }
        if c.chunk_state == .should_remove {
            destroy_chunk(&c);
            append(&to_remove, ipos);
        }
    }
    for ipos in to_remove {
        fmt.printf("removing : %v\n", ipos);
        delete_key(&chunks, ipos);
    }

}
```

### Steps to Reproduce
call Init (and add 1 agent)
call update every frame.
Move the agent around.
Then stand still a wafor everything to calm down, then see that it tries to remove a chunk every frame, but fails. 
### Failure Logs

Please include any relevant log snippets or files here.

**Comment by Kelimion** [comment]
The map is tested quite thoroughly already. You could add a check for deleted items being deleted and see if that reproduces the problem you've run into, @xzores. One way to do that would be to add integers to an array and a map both, then iterate over the array and delete the entry from the map and assert the length of the map is zero when done.

**Comment by xzores** [comment]
On second thought let me try to reduce the problem.

**Comment by Sanian-Creations** [comment]
At first I thought this was a coding error on your part but you may be onto something here actually.
Here's a more concise code snippet that I ran through a debugger which proves that at some point the map gets into a state when it's length is greater than 0, and you can loop over it to get the elements that it thinks it contains, but when you ask it if it contains those elements through a lookup, it returns false.

```odin
package main

import "core:math/rand"

main :: proc () {

    Void :: struct {}
    Key :: [3]i32
    my_map : map[Key]Void;

    to_remove : [dynamic]Key = make([dynamic]Key, 0, );
    
    iteration := 0
    total_items: u64

    for true {
        iteration += 1

        
        RANGE :: 20;
        
        r := Key{
            auto_cast rand.float32_range(-1000, 1000-RANGE), 
            auto_cast rand.float32_range(-1000, 1000-RANGE), 
            auto_cast rand.float32_range(-1000, 1000-RANGE), 
        };
        //my_map[{r.x, r.y, r.z}] = {};

        for x in r.x..<(r.x + RANGE) {
            for y in r.y..<(r.y + RANGE) {
                for z in r.z..<(r.z + RANGE) {
                    my_map[{x,y,z}] = {};
                }
            }
        }

        len_before := len(my_map)
        total_items += auto_cast len_before
        
        for key, e in &my_map {
            append(&to_remove, key);
        }
        
        for key in to_remove {
            delete_key(&my_map, key);
        }
        
        clear(&to_remove)

        len_after := len(my_map)

        if len_after > 0 {
            key: Key
            looped_over_map := false
            for k, _ in my_map {
                key = k
                looped_over_map = true
            }

            assert(looped_over_map)           // we looped over the map and got a key

            thinks_is_in_map := key in my_map
            assert(thinks_is_in_map == false) // but when asked, it will tell you this key is not in the map

            assert(false) // <- debug break here
        }
        
    }
}
```

**Comment by xzores** [comment]
Glad to hear I am not insane!
My guess would be that the bug resides inside "delete_key" and further inside "map_erase_dynamic".

**Comment by Sanian-Creations** [comment]
@Kelimion Please try running this code and tell me it loops infinitely, because what it does on my end certainly implies that there is a bug in map.

```odin
package main

import "core:math/rand"
import "core:fmt"
import "core:slice"

main :: proc () {

    N :: 350

    Nothing :: struct {}
    Key :: u16

    my_map      : map[Key]Nothing  // this one has issues
    unique_keys : [dynamic]Key     // this one we can trust

    // Even though the randomizer uses a set seed, the outcome of 
    // this program is still unpredictable, which is strange
    my_rand: rand.Rand
    rand.init(&my_rand, 1)

    retry: for {

        broke_map := false

        for !broke_map {

            // Fill the map and list with keys
            for i in 0..<N {
                key := cast(Key) rand.uint32(&my_rand)

                my_map[key] = Nothing{}

                if !slice.contains(unique_keys[:], key) { 
                    append(&unique_keys, key)
                }
            }


            // (try to) Delete all values
            for key in unique_keys {
                len_before := len(my_map)
                delete_key(&my_map, key)
                len_after := len(my_map)

                if len_before == len_after {
                    fmt.println("deleting key", key, "did not change the map length   <--")
                    broke_map = true
                }
            }

            clear(&unique_keys)
        }

        // we already broke the map, but we're looking for something greater!
        found_quantum_map := false

        fmt.println(`After "deleting all values", len(my_map) =`, len(my_map))
        for key, _ in my_map {
            key_in_map := (key in my_map)
            found_quantum_map ||= !key_in_map
            fmt.println(key, "is in the forloop, (key in my_map) =", key_in_map, !key_in_map ? "  <----- (!)":"")
        }

        if found_quantum_map { 
            fmt.println("\nPresenting The Quantum Map! It both contains and does not contain a value!!")
            return
        }

        fmt.println()
        clear(&my_map)
        clear(&unique_keys)
    }
}
```

**Comment by Kelimion** [comment]
```
W:\Odin>odin run bug.odin -file
deleting key 21835 did not change the map length   <--
After "deleting all values", len(my_map) = 1
21835 is in the forloop, (key in my_map) = true

deleting key 50970 did not change the map length   <--
After "deleting all values", len(my_map) = 1
50970 is in the forloop, (key in my_map) = true

deleting key 6066 did not change the map length   <--
deleting key 31141 did not change the map length   <--
After "deleting all values", len(my_map) = 2
31141 is in the forloop, (key in my_map) = true
6066 is in the forloop, (key in my_map) = true

deleting key 65039 did not change the map length   <--
After "deleting all values", len(my_map) = 1
65039 is in the forloop, (key in my_map) = false   <----- (!)

Presenting The Quantum Map! It both contains and does not contain a value!!
```

**Comment by jasonKercher** [comment]
This "fixes" it:

```diff
diff --git a/core/runtime/dynamic_map_internal.odin b/core/runtime/dynamic_map_internal.odin
index 05c03028..98d9bb52 100644
--- a/core/runtime/dynamic_map_internal.odin
+++ b/core/runtime/dynamic_map_internal.odin
@@ -701,10 +701,7 @@ map_erase_dynamic :: #force_inline proc "contextless" (#no_alias m: ^Raw_Map, #n
                mask := (uintptr(1)<<map_log2_cap(m^)) - 1 
                curr_index := uintptr(index)
 
-               // TODO(bill): determine a good value for this empirically
-               // if we do not implement backward shift deletion
-               PROBE_COUNT :: 8
-               for _ in 0..<PROBE_COUNT {
+               for {
                        next_index := (curr_index + 1) & mask
                        if next_index == index {
                                // looped around
```

When deleting, we coalesce tombstones for 8 iterations.  The issue is, there should be some kind of ending logic for the last iteration.  During this loop, if we hit an ideal (0 distance) or empty slot, we mark empty and quit the loop.  This is sort of the ending condition:

```odin
                        // if the next element is empty or has zero probe distance, then any lookup
                        // will always fail on the next, so we can clear both of them
                        hash := hs[next_index]
                        if map_hash_is_empty(hash) || map_probe_distance(m^, hash, next_index) == 0 { 
                                hs[curr_index] = 0 
                                return
                        } 
```

Otherwise, we push the tombstone out to the next slot.  On the last iteration, we do not execute the above block, and mark the current (index 8) as a tombstone:

```odin
                hs[curr_index] |= TOMBSTONE_MASK
```

It is possible had the loop not ended, this tombstone could become an empty slot.  Occasionally, this means we will *create* a tombstone where there would otherwise not be one and leads to duplication. I was printing out values of different slots and that looked like this (NOTE: capital letters in the hex means it's a tombstone):


![image](https://github.com/odin-lang/Odin/assets/9309879/7b2639d0-3167-48c3-aa5c-d4b6e168b141)

**Comment by xzores** [comment]
Will this edit be in the next release of Odin or is there more to be done?

**Comment by gingerBill** [comment]
@jasonKercher Please make a PR for this!!!

**Comment by Sanian-Creations** [comment]
Is there any danger to removing the limit on that forloop? Does it open up the possibility for it to loop infinitely or is that impossible? The way you say "fixed" in quotations makes me think that this is not a perfect solution.

**Comment by jasonKercher** [comment]
@Sanian-Creations Indeed, it is not *the* solution.  I was just trying to identify the source of the issue.  That deletion routine looks like a backward shift that just misses a corner case.  In reality, there should not be a backward shift (in the delete) at all because that breaks `delete_key` during iteration.  

There may yet be a simple solution.  I have to play with it some more.  If not, I have another version that does work.  When I worked on the map implementation before, I had a version that coalesced tombstones during insertion.  It worked, but the logic got more complicated than I liked.  I settled on the current version that is a bit more timid.  So worst case, I have a alternative that works assuming I can find it =]
